sw/device/lib/dif/dif_pwrmgr.c

```diff
@@ -4,8 +4,543 @@
+ */
+_Static_assert(kDifPwrmgrIrqWakeup == PWRMGR_INTR_COMMON_WAKEUP,
+               "Layout of interrupt registers changed.");
+
+/**
+ * Register information for wakeup and reset requests.
+ */
+typedef struct request_reg_info {
+  ptrdiff_t write_enable_reg_offset;
+  uint8_t write_enable_bit_index;
+  ptrdiff_t sources_reg_offset;
+  ptrdiff_t current_sources_reg_offset;
+  bitfield_field32_t bitfield;
+} request_reg_info_t;
+
+static const request_reg_info_t request_reg_infos[2] = {
+    [kDifPwrmgrReqTypeWakeup] =
+        {
+            .write_enable_reg_offset = PWRMGR_WAKEUP_EN_REGWEN_REG_OFFSET,
+            .write_enable_bit_index = PWRMGR_WAKEUP_EN_REGWEN_EN,
+            .sources_reg_offset = PWRMGR_WAKEUP_EN_REG_OFFSET,
+            .current_sources_reg_offset = PWRMGR_WAKE_STATUS_REG_OFFSET,
+            .bitfield =
+                {
+                    .mask = kDifPwrmgrWakeupRequestSourceOne,
+                    .index = 0,
+                },
+        },
+    [kDifPwrmgrReqTypeReset] =
+        {
+            .write_enable_reg_offset = PWRMGR_RESET_EN_REGWEN_REG_OFFSET,
+            .write_enable_bit_index = PWRMGR_RESET_EN_REGWEN_EN,
+            .sources_reg_offset = PWRMGR_RESET_EN_REG_OFFSET,
+            .current_sources_reg_offset = PWRMGR_RESET_STATUS_REG_OFFSET,
+            .bitfield =
+                {
+                    .mask = kDifPwrmgrResetRequestSourceOne,
+                    .index = 0,
+                },
+        },
+};
+
+/**
+ * Checks if a value is a valid `dif_pwrmgr_toggle_t`.
+ */
+DIF_WARN_UNUSED_RESULT
+static bool is_valid_toggle(dif_pwrmgr_toggle_t val) {
+  return val == kDifPwrmgrToggleEnabled || val == kDifPwrmgrToggleDisabled;
+}
+
+/**
+ * Checks if a value is a valid `dif_pwrmgr_req_type_t`.
+ */
+DIF_WARN_UNUSED_RESULT
+static bool is_valid_req_type(dif_pwrmgr_req_type_t val) {
+  return val == kDifPwrmgrReqTypeWakeup || val == kDifPwrmgrReqTypeReset;
+}
+
+/**
+ * Checks if a value is a valid `dif_pwrmgr_irq_t`.
+ */
+DIF_WARN_UNUSED_RESULT
+static bool is_valid_irq(dif_pwrmgr_irq_t val) {
+  return val >= kDifPwrmgrIrqFirst && val <= kDifPwrmgrIrqLast;
+}
+
+/**
+ * Checks if a value is valid for, i.e. fits in the mask of, a
+ * `bitfield_field32_t`.
+ */
+DIF_WARN_UNUSED_RESULT
+static bool is_valid_for_bitfield(uint32_t val, bitfield_field32_t bitfield) {
+  return (val & bitfield.mask) == val;
+}
+
+/**
+ * Checks if the control register is locked.
+ *
+ * Control register is locked when low power is enabled and WFI occurs. It is
+ * unclocked when the chip transitions back to active power state.
+ */
+DIF_WARN_UNUSED_RESULT
+static bool control_register_is_locked(const dif_pwrmgr_t *pwrmgr) {
+  if (!bitfield_bit32_read(
+          mmio_region_read32(pwrmgr->params.base_addr,
+                             PWRMGR_CTRL_CFG_REGWEN_REG_OFFSET),
+          PWRMGR_CTRL_CFG_REGWEN_EN)) {
+    return true;
+  }
+  return false;
```

Comment:
Could you drop the `if` and just return directly?

Reply:
Huh, I'm glad I didn't wrap this in another `if` :)